# Importing the libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

# Importing the dataset

In [ ]:
df = pd.read_csv('Cosmetics.csv')
df.sample
df.head(10)

Trans.   Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  \
0        1    0      1            1        1          1                0   
1        2    0      0            1        0          1                0   
2        3    0      1            0        0          1                1   
3        4    0      0            1        1          1                0   
4        5    0      1            0        0          1                0   
5        6    0      0            0        0          1                0   
6        7    0      1            1        1          1                0   
7        8    0      0            1        1          0                0   
8        9    0      0            0        0          1                0   
9       10    1      1            1        1          0                0   

   Bronzer  Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  \
0        1          1        1           0           0          0         0   
1        1          1        0           0           1          1         0   
2        1          1        1           1           1          1         1   
3        1          0        0           0           1          0         0   
4        1          1        1           1           0          1         1   
5        0          0        0           0           0          0         0   
6        1          1        1           1           1          1         0   
7        1          0        1           1           0          1         0   
8        0          0        0           0           1          0         1   
9        0          0        1           1           0          0         0   

   Eyeliner  
0         1  
1         0  
2         0  
3         1  
4         0  
5         1  
6         0  
7         0  
8         0  
9         0

# Pre-processing

In [ ]:
df = df.astype(bool)
df

Trans.     Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  \
0       True  False   True         True     True       True            False   
1       True  False  False         True    False       True            False   
2       True  False   True        False    False       True             True   
3       True  False  False         True     True       True            False   
4       True  False   True        False    False       True            False   
..       ...    ...    ...          ...      ...        ...              ...   
995     True  False  False        False    False      False            False   
996     True  False  False        False    False      False            False   
997     True  False   True         True     True       True            False   
998     True   True   True        False    False       True            False   
999     True  False  False        False    False       True            False   

     Bronzer  Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  \
0       True       True     True       False       False      False     False   
1       True       True    False       False        True       True     False   
2       True       True     True        True        True       True      True   
3       True      False    False       False        True      False     False   
4       True       True     True        True       False       True      True   
..       ...        ...      ...         ...         ...        ...       ...   
995    False      False    False       False       False      False     False   
996    False      False    False       False        True       True     False   
997    False       True    False       False        True      False      True   
998    False      False     True        True       False      False      True   
999     True       True    False        True       False      False     False   

     Eyeliner  
0        True  
1       False  
2       False  
3        True  
4       False  
..        ...  
995     False  
996     False  
997      True  
998      True  
999      True  

[1000 rows x 15 columns]

In [ ]:
df.drop('Trans. ', axis=1, inplace=True)
df

Bag  Blush  Nail Polish  Brushes  Concealer  Eyebrow Pencils  Bronzer  \
0    False   True         True     True       True            False     True   
1    False  False         True    False       True            False     True   
2    False   True        False    False       True             True     True   
3    False  False         True     True       True            False     True   
4    False   True        False    False       True            False     True   
..     ...    ...          ...      ...        ...              ...      ...   
995  False  False        False    False      False            False    False   
996  False  False        False    False      False            False    False   
997  False   True         True     True       True            False    False   
998   True   True        False    False       True            False    False   
999  False  False        False    False       True            False     True   

     Lip liner  Mascara  Eye shadow  Foundation  Lip Gloss  Lipstick  Eyeliner  
0         True     True       False       False      False     False      True  
1         True    False       False        True       True     False     False  
2         True     True        True        True       True      True     False  
3        False    False       False        True      False     False      True  
4         True     True        True       False       True      True     False  
..         ...      ...         ...         ...        ...       ...       ...  
995      False    False       False       False      False     False     False  
996      False    False       False        True       True     False     False  
997       True    False       False        True      False      True      True  
998      False     True        True       False      False      True      True  
999       True    False        True       False      False     False      True  

[1000 rows x 14 columns]

# Model construction using association rules

## Generate all the frequent itemsets with supports = 20%

In [ ]:
from mlxtend.frequent_patterns import apriori
frequent_itemsets = apriori(df, min_support=0.2, use_colnames=True)
frequent_itemsets

support                  itemsets
0     0.363                   (Blush)
1     0.280             (Nail Polish)
2     0.442               (Concealer)
3     0.279                 (Bronzer)
4     0.234               (Lip liner)
5     0.357                 (Mascara)
6     0.381              (Eye shadow)
7     0.536              (Foundation)
8     0.490               (Lip Gloss)
9     0.322                (Lipstick)
10    0.457                (Eyeliner)
11    0.220        (Concealer, Blush)
12    0.204      (Concealer, Mascara)
13    0.201   (Concealer, Eye shadow)
14    0.231   (Foundation, Concealer)
15    0.218    (Lip Gloss, Concealer)
16    0.297     (Eyeliner, Concealer)
17    0.321     (Eye shadow, Mascara)
18    0.211  (Foundation, Eye shadow)
19    0.201   (Lip Gloss, Eye shadow)
20    0.356   (Lip Gloss, Foundation)
21    0.238    (Foundation, Eyeliner)
22    0.227     (Lip Gloss, Eyeliner)

## What is the total number of frequent itemsets?

In [ ]:
len(frequent_itemsets)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


23

## Show all tha maximal frequent itemsets

In [ ]:
frequent_itemsets['Item Set Size'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets.sort_values(by='support', ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                  itemsets  Item Set Size
7     0.536              (Foundation)              1
8     0.490               (Lip Gloss)              1
10    0.457                (Eyeliner)              1
2     0.442               (Concealer)              1
6     0.381              (Eye shadow)              1
0     0.363                   (Blush)              1
5     0.357                 (Mascara)              1
20    0.356   (Lip Gloss, Foundation)              2
9     0.322                (Lipstick)              1
17    0.321     (Eye shadow, Mascara)              2
16    0.297     (Eyeliner, Concealer)              2
1     0.280             (Nail Polish)              1
3     0.279                 (Bronzer)              1
21    0.238    (Foundation, Eyeliner)              2
4     0.234               (Lip liner)              1
14    0.231   (Foundation, Concealer)              2
22    0.227     (Lip Gloss, Eyeliner)              2
11    0.220        (Concealer, Blush)              2
15    0.218    (Lip Gloss, Concealer)              2
18    0.211  (Foundation, Eye shadow)              2
12    0.204      (Concealer, Mascara)              2
19    0.201   (Lip Gloss, Eye shadow)              2
13    0.201   (Concealer, Eye shadow)              2

In [ ]:
frequent_itemsets.drop(columns='Item Set Size', inplace=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


## Generate all the association rules with minimum support = 20% and minimum confidence = 70%

In [ ]:
from mlxtend.frequent_patterns import association_rules
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents   consequents  antecedent support  consequent support  \
0  (Eye shadow)     (Mascara)               0.381               0.357   
1     (Mascara)  (Eye shadow)               0.357               0.381   
2   (Lip Gloss)  (Foundation)               0.490               0.536   

   support  confidence      lift  leverage  conviction  zhangs_metric  
0    0.321    0.842520  2.359999  0.184983    4.083050       0.930971  
1    0.321    0.899160  2.359999  0.184983    6.138417       0.896222  
2    0.356    0.726531  1.355468  0.093360    1.696716       0.514210

# Model deployment

## Show rule with highest confidence

In [ ]:
rules.sort_values(by='confidence', ascending=False)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents   consequents  antecedent support  consequent support  \
1     (Mascara)  (Eye shadow)               0.357               0.381   
0  (Eye shadow)     (Mascara)               0.381               0.357   
2   (Lip Gloss)  (Foundation)               0.490               0.536   

   support  confidence      lift  leverage  conviction  zhangs_metric  
1    0.321    0.899160  2.359999  0.184983    6.138417       0.896222  
0    0.321    0.842520  2.359999  0.184983    4.083050       0.930971  
2    0.356    0.726531  1.355468  0.093360    1.696716       0.514210

## Other Question

### Explain the meaning of this rule

A => B with confidence = C and support = D means if customers buy A, they have a chance of C% to buy B too. But customer only buy both A and B at D%

---   
Example.  
(Mascara) =>	(Eye shadow) wtih confidence = 0.899160 and support = 0.321	means if customer buy Mascara, they have a chance of 89.91% to buy Eye shadow too. But customer only buy both Mascara and Eye shadow at 32.1%

### Explain how support, confidence and lift are calculated?

Support = Freq(A, B) / N  
Confidence =  Freq(A, B) / Freq(A)  

### Show example of using these rules to increase sales.

1. Create a bundle promotion that include both items with high confidence(ex. Mascara and Eye shadow) at a discount price
2. Position items in shelf with high confidence together
